core/trino-main/src/main/java/io/trino/metadata/MetadataListing.java

```diff
@@ -83,83 +84,200 @@ public static SortedSet<String> listSchemas(Session session, Metadata metadata,
+            if (schemaName.isPresent()) {
+                // we don't use metadata.schemaExists(), because this would change semantics of the method (all vs visible schemas)
+                if (!schemaNames.contains(schemaName.get())) {
+                    return ImmutableSortedSet.of();
+                }
+                schemaNames = ImmutableSet.of(schemaName.get());
             }
-            schemaNames = ImmutableSet.of(schemaName.get());
+            return ImmutableSortedSet.copyOf(accessControl.filterSchemas(session.toSecurityContext(), catalogName, schemaNames));
+        }
+        catch (RuntimeException exception) {
+            if (exception instanceof TrinoException trinoException) {
+                throw new TrinoException(
+                        trinoException::getErrorCode,
+                        "Error listing schema names for catalog %s: %s".formatted(catalogName, exception.getMessage()),
+                        exception);
+            }
+            throw new TrinoException(
+                    StandardErrorCode.GENERIC_INTERNAL_ERROR,
+                    "Error listing schema names for catalog %s".formatted(catalogName),
+                    exception);
         }
-        return ImmutableSortedSet.copyOf(accessControl.filterSchemas(session.toSecurityContext(), catalogName, schemaNames));
     }
 
     public static Set<SchemaTableName> listTables(Session session, Metadata metadata, AccessControl accessControl, QualifiedTablePrefix prefix)
     {
-        Set<SchemaTableName> tableNames = metadata.listTables(session, prefix).stream()
-                .map(QualifiedObjectName::asSchemaTableName)
-                .collect(toImmutableSet());
+        try {
+            Set<SchemaTableName> tableNames = metadata.listTables(session, prefix).stream()
+                    .map(QualifiedObjectName::asSchemaTableName)
+                    .collect(toImmutableSet());
 
-        // Table listing operation only involves getting table names, but not any metadata. So redirected tables are not
-        // handled any differently. The target table or catalog are not involved. Thus the following filter is only called
-        // for the source catalog on source table names.
-        return accessControl.filterTables(session.toSecurityContext(), prefix.getCatalogName(), tableNames);
+            // Table listing operation only involves getting table names, but not any metadata. So redirected tables are not
+            // handled any differently. The target table or catalog are not involved. Thus the following filter is only called
+            // for the source catalog on source table names.
+            return accessControl.filterTables(session.toSecurityContext(), prefix.getCatalogName(), tableNames);
+        }
+        catch (RuntimeException exception) {
+            if (exception instanceof TrinoException trinoException) {
```

Comment:
Let's add a helper method
```java
try {
    ...
}
catch (RuntimeException e) {
    throw handleListingException(e, "tables");
}
```
Then we don't have to repeat the same error handling in many places.